### PR TITLE
Properly reject misformatted version pragmas

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: liblangutil/SemVerHandler.cpp [hunk 1]
```diff
@@ -231,6 +231,10 @@ SemVerMatchExpression::MatchComponent SemVerMatchExpressionParser::parseMatchCom
 		component.levelsPresent++;
 		if (currentChar() == '.')
 			nextChar();
+		else if (currentChar() == char(-1)) // EOF
+			break;
+		else if ('0' < currentChar() || currentChar() > '9')
```
Comment (cameel) [inline liblangutil/SemVerHandler.cpp]: I think you have an error here because as is the condition is equivalent to just `'0' < currentChar()`:
```suggestion
		else if ('0' > currentChar() || currentChar() > '9')
```

I think this might be nicer for readability because it looks a bit like a range:
```suggestion
		else if (currentChar() < '0' || '9' < currentChar())
```

File: test/libsolidity/syntaxTests/pragma/broken_version_1.sol [hunk 2]
```diff
@@ -1,3 +1,3 @@
 pragma solidity ^0^1;
```
Comment (cameel) [inline test/libsolidity/syntaxTests/pragma/broken_version_1.sol]: Could you also add a case with a character that is before `'0'` in the sort order?

File: liblangutil/SemVerHandler.cpp (hunk 1 above)
Comment (axic) [reply liblangutil/SemVerHandler.cpp]: Thanks! I think there is more work to do here. Did this before Christmas and I thought it was ready, but this logic needs to be rewritten somewhat.

**Comment by cameel** [comment]
@axic Will you be finishing this any time soon or do you want someone to take over?

**Comment by leonardoalt** [comment]
> @axic Will you be finishing this any time soon or do you want someone to take over?

Bump

**Comment by chriseth** [comment]
Bump.

File: docs/layout-of-source-files.rst [hunk 3]
```diff
@@ -79,7 +79,7 @@ be sure that your code compiles the way you intended. The exact version of the
 compiler is not fixed, so that bugfix releases are still possible.
 
 It is possible to specify more complex rules for the compiler version,
-these follow the same syntax used by `npm <https://docs.npmjs.com/cli/v6/using-npm/semver>`_.
+these follow the same syntax used by `npm <https://nodejs.dev/learn/semantic-versioning-using-npm>`_.
```
Comment (chriseth) [inline docs/layout-of-source-files.rst]: I think the first link is better, to be honest.

File: liblangutil/SemVerHandler.cpp [hunk 4]
```diff
@@ -240,8 +240,12 @@ SemVerMatchExpression::MatchComponent SemVerMatchExpressionParser::parseMatchCom
 		component.levelsPresent++;
 		if (currentChar() == '.')
 			nextChar();
-		else
+		else if (currentChar() == char(-1)) // EOF
 			break;
+		else if (("|<>=^- 0123456789"sv).find(currentChar()) != string_view::npos)
```
Comment (chriseth) [inline liblangutil/SemVerHandler.cpp]:
```suggestion
		else if (!contains("|<>=^- 0123456789"sv, currentChar())
```

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: What about tilde?

File: docs/layout-of-source-files.rst (hunk 3 above)
Comment (christianparpart) [reply docs/layout-of-source-files.rst]: for some reason that link did not work for me when I was trying to read click on it.

Comment (christianparpart) [reply docs/layout-of-source-files.rst]: ok, i must have used an older version at readthedocs maybe. will drop this commit.

File: liblangutil/SemVerHandler.cpp (hunk 4 above)
Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: thx. fixed. :)

File: test/libsolidity/syntaxTests/pragma/unterminated_version.sol [hunk 5]
```diff
@@ -1,4 +1,4 @@
 pragma solidity 0.4.3
 pragma abicoder v2;
 // ----
-// ParserError 1684: (0-41): Found version pragma, but failed to parse it. Please ensure there is a trailing semicolon.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/pragma/unterminated_version.sol]: Here, the suggestion would be quite useful. Why did you remove it?

File: liblangutil/SemVerHandler.cpp [hunk 6]
```diff
@@ -240,8 +241,12 @@ SemVerMatchExpression::MatchComponent SemVerMatchExpressionParser::parseMatchCom
 		component.levelsPresent++;
 		if (currentChar() == '.')
 			nextChar();
-		else
+		else if (currentChar() == char(-1)) // EOF
 			break;
+		else if (contains("|<>=^~- 0123456789"sv, currentChar()))
```
Comment (chriseth) [inline liblangutil/SemVerHandler.cpp]: Does this support combining components without space (`>5<7`)? Do we really want to support that?

File: test/libsolidity/syntaxTests/pragma/unterminated_version.sol (hunk 5 above)
Comment (christianparpart) [reply test/libsolidity/syntaxTests/pragma/unterminated_version.sol]: Because it's not always that reason. At least to me this reason sounds confusing when the syntax error is actually way before the `;`. I can drop that single commit though.

File: liblangutil/SemVerHandler.cpp (hunk 6 above)
Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: We do not enforce a whitespace in between the symbols as we are reusing the solidity `Scanner` for tokenization.

NB: Enforcing a  (same line) whitespace in between the version components would require columnar token information that we currently do not pass along.

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: Ok, I see, so this is fine.

What I don't understand: How can `currentChar()` be a digit? Shouldn't it be cosumed by `parseVersionPart()`?

**Comment by leonardoalt** [comment]
Needs rebase

File: liblangutil/SemVerHandler.cpp (hunk 6 above)
Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: The tokens are already pre-tokenized. `parseVersionPart()` is using `nextChar()` to consume more chars as long as it's a digit but stops when `m_pos` points to the next token in the the array of tokens. So the next call to `currentChar()` outside `parseVersionPart()` already points to the first char of the next token.

It's weird, I know. :)

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: So you are saying this might be the case for `>=0.1 1`, right? Can you check if we have a test for that and that it parses as ">=0.1.0 && =1.*.*" instead of `>=0.11`?

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: On the other hand, do we really want to list all the allowed characters here? What if we just break in all cases (and not throw SemVerError)? Will we get an error later?

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: I mean does it make sense to check for EOF one level higher in the parser?

Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: > On the other hand, do we really want to list all the allowed characters here? What if we just break in all cases (and not throw SemVerError)? Will we get an error later?

I would have preferred that initially. Tried it, failed, I think that would require much more code changes. If we want to do that, I'd suggest to do that refactoring exclusively in its own PR ?

I've added the test-suggestion from above though (works).

File: test/libsolidity/SemVerMatcher.cpp [hunk 7]
```diff
@@ -78,6 +81,7 @@ BOOST_AUTO_TEST_CASE(positive_range)
 		{"1.0.0", "1.0.0"},
 		{"1.0", "1.0.0"},
 		{"1", "1.0.0"},
+		{">=1.2.3 1.2.3", "1.2.3"},
```
Comment (cameel) [inline test/libsolidity/SemVerMatcher.cpp]: This case is a bit out of place here. I'd move it next to `>=1.2.3 >=1.2.1`.

Also, I think we should have a negative test for `">=1.2.3 1.2.3", "1.2.4"`

File: Changelog.md [hunk 8]
```diff
@@ -14,7 +14,7 @@ Bugfixes:
  * Commandline Interface: Fix extra newline character being appended to sources passed through standard input, affecting their hashes.
  * SMTChecker: Fix internal error in magic type access (``block``, ``msg``, ``tx``).
  * TypeChecker: Fix internal error when using user defined value types in public library functions.
-
+ * Parser: Properly reject misformatted version pragmas (e.g. ``^0^1`` was misparsed).
```
Comment (cameel) [inline Changelog.md]: The PR does not change the error message in `broken_version_1.sol`, which contains the test case for `^0^1`. Is this case actually affected?

File: liblangutil/SemVerHandler.cpp (hunk 6 above)
Comment (cameel) [inline liblangutil/SemVerHandler.cpp]: `*` is not listed. Stuff like `*.8.9` seems to work currently and matches e.g. both `0.8.9` and `0.7.9` but not `0.8.8` or `0.7.7`. We're missing a test case for this.

Same for `"`, `\t`, `\n`, `X` and `x`. But WTF, why is `x` a valid version?

Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: By the way it's interesting that even weird things like these actually work correctly:
- `pragma solidity >*.*.2`
- `pragma solidity ~*.*.2`
- `pragma solidity ^*.*.2`
- `pragma solidity < x`
- `pragma solidity ~"*"`
- `pragma solidity 1.2.x`

and you can do even pathological stuff like this:
- `pragma solidity x.x.x.x.x.x`
- `pragma solidity XxXxXxX`
- `pragma solidity X <xx> X`
- `pragma solidity ***"666"=xXx="666"***`
- `pragma solidity XXX = 0.8.9 = XXX` (equivalent to just `=0.8.9`)

I think that the coverage for the semver syntax is generally too sparse. If we support these weird cases we absolutely need to have them covered or they'll just break at some point and we won't notice.

Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: We should really document this stuff: #4642.

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: Can we simply fix this specific case for now? My guess is that the problem is that we have redundancy here: This condition here should mirror a condition somewhere else, but it's probably better to not mirror it at all.

File: Changelog.md (hunk 8 above)
Comment (christianparpart) [reply Changelog.md]: `^0^1` could not be *properly* parsed before. Now it does. `^0 ^1` does not make sense but is syntactically correct, as of semver spec. It is a match that can never be true though. Not sure there should be any special message for that.

File: liblangutil/SemVerHandler.cpp (hunk 6 above)
Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: > Same for ", \t, \n, X and x. But WTF, why is x a valid version?

\t \n should never make it in, nor should  a single " because we're using the solidity `Scanner` that is pre-tokenizing. @cameel

Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: I've eliminated all of these cases (rewrote some pre-existing `default:`-case) and restricted `x`-handling only a little.

Fully fixing `x` is a good question. What is `x` anyways (it resembles `UINT_MAX` and could be used for things like `0.8.x` or equivalently `0.8.*`. I did not find any use of the `UINT_MAX` meaning. So I think with the next major release we could instead just drop the special case handling of `x` and `*` (they are not part of current semver spec).

**Comment by chriseth** [comment]
`*` and `x` are part of the grammar at https://www.npmjs.com/package/semver

File: liblangutil/SemVerHandler.cpp (hunk 6 above)
Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: So what about this one? I thought you managed to get rid of the character whitelisting.

Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: > Can we simply fix this specific case for now?

I mean, documenting is here is probably not necessary but without good test coverage it's hard tell with certainty that the PR does not break stuff so I think it's pretty important.

> \t \n should never make it in, nor should a single " because we're using the solidity `Scanner` that is pre-tokenizing. 

If that's the case then won't numbers not make it either? They do have a designated token. Or is it because `Scanner` removes the `"` from the string it scans?

BTW, I mentioned `\t` mostly because I saw in test cases. If we can't get it here then great. But we should have a test case proving that.

> So I think with the next major release we could instead just drop the special case handling

Looks like we can't drop it. We want to be compatible with npm and npm does support it.

File: liblangutil/SemVerHandler.cpp [hunk 9]
```diff
@@ -252,9 +261,10 @@ unsigned SemVerMatchExpressionParser::parseVersionPart()
 {
 	auto startPos = m_pos;
 	char c = currentChar();
+	bool const isLastCharInToken = m_posInside + 1 == currentTokenLiteral().size();
```
Comment (cameel) [inline liblangutil/SemVerHandler.cpp]:
```suggestion
	bool const isLastCharInToken = (m_posInside + 1 == currentTokenLiteral().size());
```

File: liblangutil/SemVerHandler.cpp [hunk 10]
```diff
@@ -252,9 +261,10 @@ unsigned SemVerMatchExpressionParser::parseVersionPart()
...
 	nextChar();
-	if (c == 'x' || c == 'X' || c == '*')
-		return unsigned(-1);
+	if (isLastCharInToken && (c == 'x' || c == 'X' || c == '*'))
```
Comment (cameel) [inline liblangutil/SemVerHandler.cpp]: So this disallows stuff like `xxx`, `***` or `1.***.3`? Can you add some test cases for this?

File: Changelog.md (hunk 8 above)
Comment (cameel) [reply Changelog.md]: How was it parsed before? As just `^0`? Then we should be able to write a test case for it. I.e. one where `0.1.2` previously matched `^0^1` and now it does not.

File: liblangutil/SemVerHandler.cpp (hunk 10 above)
Comment (christianparpart) [reply liblangutil/SemVerHandler.cpp]: @cameel FYI: we agreed to not take care of all these invalid cases as of now but make sure that `^0^1` won't match plus some more tests.

**Review by chriseth — APPROVED** [review]

File: liblangutil/SemVerHandler.cpp (hunk 10 above)
Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: ok, but maybe you could create an issue for someone to do this eventually? I think that increasing the coverage of the pragma syntax would be a good `good first issue` for Hacktoberfest.

Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: Actually, I'm a bit confused about what happened here :) The PR now only adds a few test cases and does not change the implementation at all.

Comment (chriseth) [reply liblangutil/SemVerHandler.cpp]: yeah sorry, we discussed it in the office and could not really see what the main purpose of the PR was after checking that `^0^1` is properly parsed to `>= 0.0.0 && <1.0.0 && >=1.0.0 && <2.0.0`.

I don't think it makes sense to require spaces between the components.

Comment (cameel) [reply liblangutil/SemVerHandler.cpp]: ah, ok :) so I was right about https://github.com/ethereum/solidity/pull/11295#discussion_r725036533 then :)
